src/GraphQL.ApiTests/GraphQL.approved.txt

```diff
@@ -2134,6 +2135,8 @@ namespace GraphQL.Types
     public class SchemaTypes : System.Collections.Generic.IEnumerable<GraphQL.Types.IGraphType>, System.Collections.IEnumerable
     {
         public int Count { get; }
+        public void ApplyMiddleware(GraphQL.Instrumentation.IFieldMiddlewareBuilder fieldMiddlewareBuilder, GraphQL.Types.ISchema schema) { }
+        public void ApplyMiddleware(System.Func<GraphQL.Types.ISchema, GraphQL.Instrumentation.FieldMiddlewareDelegate, GraphQL.Instrumentation.FieldMiddlewareDelegate> transform, GraphQL.Types.ISchema schema) { }
```

Comment:
I really don't like the fact that the schema has to passed into `SchemaTypes` to configure the middleware, since logically `SchemaTypes` is a 'child' class of the schema.  It seems as though `ApplyMiddleware` would ideally be an extension method.  However, as an extension method of `ISchema`, it would need access to `AllTypes`, which would pose a "deadlock" during initialization as seen before, unless we create our own lock within `Schema` and remove `Lazy`.

In the end I left it within `SchemaTypes` anyway.

Now that I think about it, this still may not work, since the middleware delegate may attempt to access `AllTypes` while configuring the resolvers.  I will write a test for this and see if it fails.

Reply:
> middleware delegate may attempt to access AllTypes while configuring the resolvers

No as `delegate` not executed itself.

Reply:
> I really don't like the fact that the schema has to passed into `SchemaTypes` to configure the middleware, since logically `SchemaTypes` is a 'child' class of the schema. 

I understand your feelings :) In fact, there is nothing wrong with that, but we can improve the situation. `schema` parameter is only used for one(ok, two) overload - ` public static IFieldMiddlewareBuilder Use<T>(this IFieldMiddlewareBuilder builder) where T : IFieldMiddleware => Use(builder, typeof(T));`/`  public static IFieldMiddlewareBuilder Use(this IFieldMiddlewareBuilder builder, System.Type middleware)` . We _can_ get rid of it now.

Reply:
This parameter is optional, the calling code can take it from the closure.

Reply:
```c#
 options.FieldMiddleware
                        .Use<CountFieldMiddleware>()
                        .Use<InstrumentFieldsMiddleware>();
```
->
```c#
 options.FieldMiddleware
                        .Use(someserviceProvider.GetService<CountFieldMiddleware>())
                        .Use(new InstrumentFieldsMiddleware());
```

Reply:
I don't think we should eliminate the `Use<MyMiddleware>` syntax if we can help it.

Reply:
we could write an extension method on Schema, if we want....not ideal though.....

```cs
schema.UseMiddlware<CountFieldMiddleware>();
```

which is an extension method for:

```cs
schema.FieldMiddleware.Use(schema.ServiceProvider.GetRequiredService<CountFieldMiddleware>());
```

Reply:
50/50 Let's leave it as it is. In some next PR we will think about it better.

Reply:
Removed. OK.